draft-ietf-quic-qpack.md

```diff
@@ -619,14 +619,22 @@ QPACK defines two unidirectional stream types:
    It carries an unframed sequence of decoder instructions from decoder
    to encoder.
 
-<!-- s/exactly/no more than/  ? -->
 HTTP/3 endpoints contain a QPACK encoder and decoder. Each endpoint MUST
-initiate a single encoder stream and decoder stream. Receipt of a second
-instance of either stream type be MUST treated as a connection error of type
+initiate at most one encoder stream and one decoder stream. Receipt of a second
+instance of either stream type MUST be treated as a connection error of type
 HTTP_WRONG_STREAM_COUNT. These streams MUST NOT be closed. Closure of either
 unidirectional stream type MUST be treated as a connection error of type
 HTTP_CLOSED_CRITICAL_STREAM.
 
+An endpoint MAY avoid creating its own encoder stream if the maximum size of
+the dynamic table permitted by the peer is zero.
+
+An endpoint MAY avoid creating its own decoder stream if the maximum size of
+its own dynamic table is zero.
+
+An endpoint MUST allow its peer to create both encoder and decoder streams
```

Comment:
> An endpoint MUST allow its peer to create both encoder and decoder streams

Why?

One way to prevent the peer from creating QPACK streams is to set the limit of unidirectional streams to 1 in the transport parameters.

Reply:
Going back over some of the discussion thread on this topic, was the preference to do lazy loading of streams in response to SETTINGS? In which case, when an endpoint receives a SETTINGS_QPACK_MAX_TABLE_CAPACITY it should send a QUIC MAX_STREAMS to allow the peer to open the encoder stream, and hope that the peer has also given enough streams to open the decoder stream.

Its hard to tell from the different threads what the consensus was.

Reply:
Hmmm, I think I was reading https://github.com/quicwg/base-drafts/issues/2100#issuecomment-459241694 wrong.

But yeah, the intention there was to disallow setting the uni streams limit to less than 3 in order to allow the other side to open all of their streams. This is effectively what the spec says right now I think, since each side MUST create their streams and if the peer's limits prevent that, than the endpoint can't create them.

I would also be fine aloowing one side to prevent the other from opening and making that not an error, but either way I think we should specifiy what to do since the spec right now is far from clear on this.

Reply:
> Hmmm, I think I was reading #2100 (comment) wrong.
> 
> But yeah, the **intention there was to disallow setting the uni streams limit to less than 3** in order to allow the other side to open all of their streams.

(Emphasis added)

I see how you get that from that comment by @MikeBishop:

> Discussed in Tokyo; do not require creating the streams, but creating them is not forbidden even if the setting prevents their use

But the "setting" refers to the QPACK settings -- namely, the size of the dynamic table -- and not to the stream limits.  Otherwise, how could it not be forbidden to create streams over the stream number limits specified by the peer?

I agree that more clarity is required.  My preference would be to allow the unidirectional stream limit to be set to 1 to allow just the HTTP Control Stream.

Reply:
> Both clients and servers SHOULD send a value of three or greater for the QUIC transport parameter initial_max_uni_streams.

I think the current consensus regarding QPACK streams is that you cannot prevent their creation, so clients and servers MUST give 3 or more.  If we allow the peer to limit to one unidirectional stream, there's potentially nothing from stopping an implementation from using that to create an encoder or decoder stream.

Reply:
> I think the current consensus regarding QPACK streams is that you cannot prevent their creation

What does that mean -- "[one] cannot prevent their creation?"  Surely, setting the limit low does just that.

The general MO of this WG has been to assume that the implementers are capable of making sensible choices.  Creating an unnecessary QPACK stream instead of the HTTP Control Stream when the allowance is just one stream is far from a subtle "gotcha."

Reply:
The trap here is that a peer sets initial uni directional streams. And then may increase this value. 

Uni directional stream accounting needs to factor in qpack, push promise and whatever other extension mechanisms might come to pass.
The explicit signal that's missing is for QPACK off. If you dont tell the peer you wont use QPACK, and it advertised a size > 0. It will have to entertain that this could happen at any point in the connection. 

In a pathological case, you give me 3 credits and I use them to send control and 2 grease streams. What happens next?

I think these sorts of issues will come to light, the specs are not so helpful in guidance right now.

Reply:
@dtikhonov : I mean to say, that an implementation that prevents the creation of QPACK streams is not conformant, in the same way it would not be conformant to set it to 0 and block the creation of the control stream.  This is my understanding of the consensus based on the last discussion in Toyko.

@LPardue : I'm not sure that's such a big deal.  If you planned to use the dynamic table, but wasted your stream on grease instead of an encoder stream, then you can't use the table.  If you disabled the dynamic table for your peer, then you don't need to create the decoder stream anyways.  If you allowed the peer to use the dynamic table but can't create a decoder stream, then your compression will eventually degrade to ~static only but I think the connection will still work.

Reply:
>  I'm not sure that's such a big deal. If you planned to use the dynamic table, but wasted your stream on grease instead of an encoder stream, then you can't use the table.

The problem is that the sender has no intention of creating an encoder stream but the receiver doesn't know that. If the receiver is eager to use QPACK, it keeps giving back a stream credit for every unidirectional stream that is "wasted" before creating the encoder stream. Even if it didn't give back credit immediately and did let the max uni stream limit be hit, there is no guarantee that the next time the receiver gives credit (say for a push promise) that then it wouldn't spontaneously be used for QPACK.

Reply:
This sounds like a case of "don't do that".  The sender gets to choose to how to use the uni stream credit it receives, and I don't think receivers should overthink this.  A receiver should choose the value based on how many concurrent uni streams they can handle, since there's no way to guaranteer the sender will use them for any particular type.  If the sender wants to use the dynamic table at any point in the connection, maybe we should recommend they open the stream as soon as possible.

Reply:
Thinking about this a little more, its a nice feature of HPACK that it can chose when to start using dynamic compression, with little overhead. 

My understanding though is that we can't use MAX_STREAMS to contorl concurrency like you describe (I'm probably wrong here). But from the transport document section 19.11

> Note that these frames (and the corresponding transport parameters)
   do not describe the number of streams that can be opened
   concurrently.  The limit includes streams that have been closed as
   well as those that are open.

Reply:
I think most implementations will take closed streams into account to know when to issue more uni stream credit.  In the case of servers, there's never a need to increase it in vanilla HTTP/3.   Clients need only increase it when they want more pushes (or grease, or extensions, etc).

> Thinking about this a little more, its a nice feature of HPACK that it can chose when to start using dynamic compression, with little overhead.

I suppose that's a feature of HPACK, but is it really all that useful?  I think of it the other way - HPACK was simpler to setup (didn't require any extra streams) but that's because the underlying transport provides the strong ordering guarantee.  If you want QPACK dynamic table then you have to open some streams.  Just do it at the beginning of the connection when you know you have credit, or risk not being able to use it later, if you consume the credit with some other stream.

Reply:
> I think most implementations will take closed streams into account to know when to issue more uni stream credit. In the case of servers, there's never a need to increase it in vanilla HTTP/3. Clients need only increase it when they want more pushes (or grease, or extensions, etc).

Once the three basic streams are opened, servers never need to increase the number of streams, true.  However, the server still needs to keep giving credit as streams close until those three are opened at least.

Reply:
> However, the server still needs to keep giving credit as streams close until those three are opened at least.

There's a difference between "must give >= 3 initial uni credits" and "must always have enough credit to open the control+qpack streams".  Do we have consensus on what the requirement should be?  I'm inclined to say it's the former.  Don't burn your credits on grease/push/extensions if you're going to need the other streams.